drivers/usb/device/usb_dc_stm32.c

```diff
@@ -740,6 +740,14 @@ int usb_dc_ep_clear_stall(const uint8_t ep)
 		return -EIO;
 	}
 
+	/* The DataInCallback will never be called at this point for any pending
+	 * transaction associated with this endpoint as it's status has been set to NAK.
+	 * Reset the IN semaphore to prevent perpetual locked state.
+	 */
+	if (USB_EP_DIR_IS_IN(ep) && ep_state->ep_type != EP_TYPE_ISOC) {
+		k_sem_give(&ep_state->write_sem);
+	}
+
```

Comment:
Instead of giving up, the driver could restart the current transfer. It is still unclear from your report what is going on with your device and host. There could be a check `if(!ep_state->ep_stalled)` to not call HAL_PCD_EP_ClrStall() if the Clear Feature request is spurious.